AutoDefineAddon/libs/cardbuilder.py

```diff
@@ -0,0 +1,367 @@
+                valid_entries.append(entry)
+    return valid_entries
+
+
+ValidAndPotentialEntries = namedtuple('Entries', ['valid', 'potential'])
+
+
+def get_entries_from_api(word, url):
+    if "YOUR_KEY_HERE" in url:
+        return []
+    try:
+        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:62.0)'
+                                                                 ' Gecko/20100101 Firefox/62.0'})
+        returned = urllib.request.urlopen(req).read()
+        if "Invalid API key" in returned.decode("UTF-8"):
+            showInfo("API key '%s' is invalid. Please double-check you are using the key labeled \"Key (Dictionary)\". "
+                     "A web browser with the web page that lists your keys will open." % url.split("?key=")[1])
+            webbrowser.open("https://www.dictionaryapi.com/account/my-keys.htm")
+            return []
+        if "Results not found" in returned.decode("UTF-8"):
+            return []
+        if url.find("xml") != -1:
+            etree = ET.fromstring(returned)
+            return etree.findall("entry")
+        elif url.find("json") != -1:
+            content = json.loads(returned)
+            return content
+    except URLError:
+        return []
+    except ValueError:
+        return []
+    except (ET.ParseError, RemoteDisconnected):
+        showInfo("Couldn't parse API response for word '%s'. "
+                 "Please submit an issue to the AutoDefine GitHub (a web browser window will open)." % word)
+        webbrowser.open("https://github.com/z1lc/AutoDefine/issues/new?title=Parse error for word '%s'"
+                        "&body=Anki Version: %s%%0APlatform: %s %s%%0AURL: %s%%0AStack Trace: %s"
+                        % (word, version, platform.system(), platform.release(), url, traceback.format_exc()), 0, False)
+class Card():
+    entries = []
+    fields = ["", "", "", ""]
+    def parse(self, entries) -> list:
+        pass
+
+    def getField(self, id):
+        # FIXME: out of range
+        return self.fields[id]
+
+    def serialize(self, editor):
+        if editor.note:
+            i = 0
+            for field in self.fields:
+                if field.find("wav") != -1:
+                    editor.note.fields[i] = editor.urlToLink(field)
+                else:
+                    editor.note.fields[i] = field
+                editor.loadNote()
+                i += 1
+
+class CardBuilder:
+    _card = Card()
+    def __init__(self, word):
+        self._card.fields[0] = word
+
+    def addDefinition(self):
+        self._card.fields[settings.DEFINITION_FIELD] = ""
+
+    def addTranscription(self):
+        self._card.fields[settings.PHONETIC_TRANSCRIPTION_FIELD] = ""
+
+    def addPronunciation(self):
+        self._card.fields[settings.PRONUNCIATION_FIELD] = ""
+
+    def getCard(self) -> Card:
+        return self._card
+
+class CollegiateCardBuilder(CardBuilder):
```

Comment:
Looks like this is functioning as a card builder for both medical as well as collegiate? Let's rename this as GenericCardBuilder or something and then subclass Collegiate and Medical from it.